<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="bioMONAI core functions">

<title>bioMONAI - Core</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="bioMONAI - Core">
<meta property="og:description" content="bioMONAI core functions">
<meta property="og:image" content="https://deepCLEM.github.io/bioMONAI/00_core_files/figure-html/cell-14-output-4.png">
<meta property="og:site-name" content="bioMONAI">
<meta property="og:image:height" content="351">
<meta property="og:image:width" content="516">
<meta name="twitter:title" content="bioMONAI - Core">
<meta name="twitter:description" content="bioMONAI core functions">
<meta name="twitter:image" content="https://deepCLEM.github.io/bioMONAI/00_core_files/figure-html/cell-14-output-4.png">
<meta name="twitter:image-height" content="351">
<meta name="twitter:image-width" content="516">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">bioMONAI</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://deepclem.github.io" rel="" target="">
 <span class="menu-text">Blog</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-resources" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Resources</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-resources">    
        <li>
    <a class="dropdown-item" href="https://monai.io/" rel="" target=""><i class="bi bi-link" role="img">
</i> 
 <span class="dropdown-text">monai</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://docs.fast.ai" rel="" target=""><i class="bi bi-link" role="img">
</i> 
 <span class="dropdown-text">fast.ai</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://fastcore.fast.ai" rel="" target=""><i class="bi bi-link" role="img">
</i> 
 <span class="dropdown-text">fastcore</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://fastxtend.benjaminwarner.dev" rel="" target=""><i class="bi bi-link" role="img">
</i> 
 <span class="dropdown-text">fastxtend</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://nbdev.fast.ai" rel="" target=""><i class="bi bi-link" role="img">
</i> 
 <span class="dropdown-text">nbdev</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://github.com/deepCLEM/bioMONAI/issues" rel="" target=""><i class="bi bi-bug" role="img">
</i> 
 <span class="dropdown-text">Report an Issue</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/deepCLEM/bioMONAI/" rel="" target=""><i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/" rel="" target=""><i class="bi bi-twitter" role="img" aria-label="Twitter">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./core.html">Core</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">bioMONAI</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./core.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Core</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./io.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">I/O</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./losses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Loss functions</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./nets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Networks</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./transforms.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Transforms</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./metrics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Metrics</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./callbacks.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Callbacks</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Datasets</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./visualize.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Visualize</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Tutorials</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Tutorials/tutorial_classification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Image Classification 2D</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Tutorials/tutorial_classification3d.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Image Classification 3D</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Tutorials/tutorial_denoising.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Denoising</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Tutorials/tutorial_multispectral_classification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Multispectral Classification</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#imports" id="toc-imports" class="nav-link active" data-scroll-target="#imports">Imports</a>
  <ul>
  <li><a href="#datablock" id="toc-datablock" class="nav-link" data-scroll-target="#datablock">DataBlock</a></li>
  <li><a href="#dataloaders" id="toc-dataloaders" class="nav-link" data-scroll-target="#dataloaders">DataLoaders</a></li>
  <li><a href="#learner" id="toc-learner" class="nav-link" data-scroll-target="#learner">Learner</a></li>
  <li><a href="#showgraphcallback" id="toc-showgraphcallback" class="nav-link" data-scroll-target="#showgraphcallback">ShowGraphCallback</a></li>
  <li><a href="#csvlogger" id="toc-csvlogger" class="nav-link" data-scroll-target="#csvlogger">CSVLogger</a></li>
  <li><a href="#cells3d" id="toc-cells3d" class="nav-link" data-scroll-target="#cells3d">cells3d</a></li>
  </ul></li>
  <li><a href="#engine" id="toc-engine" class="nav-link" data-scroll-target="#engine">Engine</a>
  <ul>
  <li><a href="#read_yaml" id="toc-read_yaml" class="nav-link" data-scroll-target="#read_yaml">read_yaml</a></li>
  <li><a href="#dictlist_to_funclist" id="toc-dictlist_to_funclist" class="nav-link" data-scroll-target="#dictlist_to_funclist">dictlist_to_funclist</a></li>
  <li><a href="#fasttrainer" id="toc-fasttrainer" class="nav-link" data-scroll-target="#fasttrainer">fastTrainer</a>
  <ul class="collapse">
  <li><a href="#example-train-a-model-with-configuration-from-a-yaml-file." id="toc-example-train-a-model-with-configuration-from-a-yaml-file." class="nav-link" data-scroll-target="#example-train-a-model-with-configuration-from-a-yaml-file.">Example: train a model with configuration from a YAML file.</a></li>
  </ul></li>
  <li><a href="#visiontrainer" id="toc-visiontrainer" class="nav-link" data-scroll-target="#visiontrainer">visionTrainer</a></li>
  </ul></li>
  <li><a href="#evaluation" id="toc-evaluation" class="nav-link" data-scroll-target="#evaluation">Evaluation</a>
  <ul>
  <li><a href="#display_statistics_table" id="toc-display_statistics_table" class="nav-link" data-scroll-target="#display_statistics_table">display_statistics_table</a></li>
  <li><a href="#plot_histogram_and_kde" id="toc-plot_histogram_and_kde" class="nav-link" data-scroll-target="#plot_histogram_and_kde">plot_histogram_and_kde</a></li>
  <li><a href="#format_sig" id="toc-format_sig" class="nav-link" data-scroll-target="#format_sig">format_sig</a></li>
  <li><a href="#calculate_statistics" id="toc-calculate_statistics" class="nav-link" data-scroll-target="#calculate_statistics">calculate_statistics</a></li>
  <li><a href="#compute_metric" id="toc-compute_metric" class="nav-link" data-scroll-target="#compute_metric">compute_metric</a></li>
  <li><a href="#compute_losses" id="toc-compute_losses" class="nav-link" data-scroll-target="#compute_losses">compute_losses</a></li>
  <li><a href="#evaluate_model" id="toc-evaluate_model" class="nav-link" data-scroll-target="#evaluate_model">evaluate_model</a></li>
  <li><a href="#evaluate_classification_model" id="toc-evaluate_classification_model" class="nav-link" data-scroll-target="#evaluate_classification_model">evaluate_classification_model</a></li>
  </ul></li>
  <li><a href="#utils" id="toc-utils" class="nav-link" data-scroll-target="#utils">Utils</a>
  <ul>
  <li><a href="#attributesfromdict" id="toc-attributesfromdict" class="nav-link" data-scroll-target="#attributesfromdict">attributesFromDict</a></li>
  <li><a href="#get_device" id="toc-get_device" class="nav-link" data-scroll-target="#get_device">get_device</a></li>
  <li><a href="#img2float" id="toc-img2float" class="nav-link" data-scroll-target="#img2float">img2float</a></li>
  <li><a href="#img2tensor" id="toc-img2tensor" class="nav-link" data-scroll-target="#img2tensor">img2Tensor</a></li>
  <li><a href="#apply_transforms" id="toc-apply_transforms" class="nav-link" data-scroll-target="#apply_transforms">apply_transforms</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/deepCLEM/bioMONAI/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Core</h1>
</div>

<div>
  <div class="description">
    bioMONAI core functions
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>
<script src="https://static.cloudflareinsights.com/beacon.min.js?token=62f27bdcb8964407a52676798db9edfa" defer=""></script>
<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<section id="imports" class="level2">
<h2 class="anchored" data-anchor-id="imports">Imports</h2>
<p>This section includes essential imports used throughout the core library, providing foundational tools for data handling, model training, and evaluation. Key imports cover areas such as data blocks, data loaders, custom loss functions, optimizers, callbacks, and logging.</p>
<hr>
<p><a href="https://github.com/fastai/fastai/blob/master/fastai/data/block.py#LNone" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="datablock" class="level3">
<h3 class="anchored" data-anchor-id="datablock">DataBlock</h3>
<blockquote class="blockquote">
<pre><code> DataBlock (blocks:list=None, dl_type:TfmdDL=None, getters:list=None,
            n_inp:int=None, item_tfms:list=None, batch_tfms:list=None,
            get_items=None, splitter=None, get_y=None, get_x=None)</code></pre>
</blockquote>
<p><em>Generic container to quickly build <code>Datasets</code> and <code>DataLoaders</code>.</em></p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>blocks</td>
<td>list</td>
<td>None</td>
<td>One or more <code>TransformBlock</code>s</td>
</tr>
<tr class="even">
<td>dl_type</td>
<td>TfmdDL</td>
<td>None</td>
<td>Task specific <code>TfmdDL</code>, defaults to <code>block</code>’s dl_type or<code>TfmdDL</code></td>
</tr>
<tr class="odd">
<td>getters</td>
<td>list</td>
<td>None</td>
<td>Getter functions applied to results of <code>get_items</code></td>
</tr>
<tr class="even">
<td>n_inp</td>
<td>int</td>
<td>None</td>
<td>Number of inputs</td>
</tr>
<tr class="odd">
<td>item_tfms</td>
<td>list</td>
<td>None</td>
<td><code>ItemTransform</code>s, applied on an item</td>
</tr>
<tr class="even">
<td>batch_tfms</td>
<td>list</td>
<td>None</td>
<td><code>Transform</code>s or <code>RandTransform</code>s, applied by batch</td>
</tr>
<tr class="odd">
<td>get_items</td>
<td>NoneType</td>
<td>None</td>
<td></td>
</tr>
<tr class="even">
<td>splitter</td>
<td>NoneType</td>
<td>None</td>
<td></td>
</tr>
<tr class="odd">
<td>get_y</td>
<td>NoneType</td>
<td>None</td>
<td></td>
</tr>
<tr class="even">
<td>get_x</td>
<td>NoneType</td>
<td>None</td>
<td></td>
</tr>
</tbody>
</table>
<p>The <code>DataBlock</code> class Datablock comes from the fastai library and builds datasets and dataloaders from blocks, acting as a container for creating data processing pipelines, allowing easy customization of datasets and data loaders. It enables the definition of item transformations, batch transformations, and dataset split methods, streamlining data preprocessing and loading across various stages of model training.</p>
<hr>
<p><a href="https://github.com/fastai/fastai/blob/master/fastai/data/core.py#LNone" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="dataloaders" class="level3">
<h3 class="anchored" data-anchor-id="dataloaders">DataLoaders</h3>
<blockquote class="blockquote">
<pre><code> DataLoaders (*loaders, path:str|pathlib.Path='.', device=None)</code></pre>
</blockquote>
<p><em>Basic wrapper around several <code>DataLoader</code>s.</em></p>
<p>The <code>DataLoaders</code> class is a container for managing training and validation datasets. This class wraps one or more <code>DataLoader</code> instances, ensuring seamless data management and transfer across devices (CPU or GPU) for efficient training and evaluation.</p>
<hr>
<p><a href="https://github.com/fastai/fastai/blob/master/fastai/learner.py#LNone" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="learner" class="level3">
<h3 class="anchored" data-anchor-id="learner">Learner</h3>
<p><em>Group together a <code>model</code>, some <code>dls</code> and a <code>loss_func</code> to handle training</em></p>
<p>The <code>Learner</code> class is the main interface for training machine learning models, encapsulating the model, data, loss function, optimizer, and training metrics. It simplifies the training process by providing built-in functionality for model evaluation, hyperparameter tuning, and training loop customization, allowing you to focus on model optimization.</p>
<hr>
<p><a href="https://github.com/fastai/fastai/blob/master/fastai/callback/progress.py#LNone" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="showgraphcallback" class="level3">
<h3 class="anchored" data-anchor-id="showgraphcallback">ShowGraphCallback</h3>
<blockquote class="blockquote">
<pre><code> ShowGraphCallback (after_create=None, before_fit=None, before_epoch=None,
                    before_train=None, before_batch=None, after_pred=None,
                    after_loss=None, before_backward=None,
                    after_cancel_backward=None, after_backward=None,
                    before_step=None, after_cancel_step=None,
                    after_step=None, after_cancel_batch=None,
                    after_batch=None, after_cancel_train=None,
                    after_train=None, before_validate=None,
                    after_cancel_validate=None, after_validate=None,
                    after_cancel_epoch=None, after_epoch=None,
                    after_cancel_fit=None, after_fit=None)</code></pre>
</blockquote>
<p><em>Update a graph of training and validation loss</em></p>
<p>The <code>ShowGraphCallback</code> is a convenient callback for visualizing training progress. By plotting the training and validation loss, it helps users monitor convergence and performance, making it easy to assess if the model requires adjustments in learning rate, architecture, or data handling.</p>
<hr>
<p><a href="https://github.com/fastai/fastai/blob/master/fastai/callback/progress.py#LNone" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="csvlogger" class="level3">
<h3 class="anchored" data-anchor-id="csvlogger">CSVLogger</h3>
<blockquote class="blockquote">
<pre><code> CSVLogger (fname='history.csv', append=False)</code></pre>
</blockquote>
<p><em>Log the results displayed in <code>learn.path/fname</code></em></p>
<p>The <code>CSVLogger</code> is a tool for logging model training metrics to a CSV file, offering a permanent record of training history. This feature is especially useful for long-term experiments and fine-tuning, allowing you to track and analyze model performance over time.</p>
<hr>
</section>
<section id="cells3d" class="level3">
<h3 class="anchored" data-anchor-id="cells3d">cells3d</h3>
<blockquote class="blockquote">
<pre><code> cells3d ()</code></pre>
</blockquote>
<p>*3D fluorescence microscopy image of cells.</p>
<p>The returned data is a 3D multichannel array with dimensions provided in <code>(z, c, y, x)</code> order. Each voxel has a size of <code>(0.29 0.26 0.26)</code> micrometer. Channel 0 contains cell membranes, channel 1 contains nuclei.*</p>
<p>The <code>cells3d</code> function returns a sample 3D fluorescence microscopy image. This is a valuable test image for demonstration and analysis, consisting of both cell membrane and nucleus channels. It can serve as a default dataset for evaluating and benchmarking new models and transformations.</p>
<p>The dataset has the following dimensions: <em>(60,2,256,256)</em>, which are in a <em>(z,c,y,x)</em> order: - z: 60 slices in the image - c: 2 channels where channel 0 represents the cell membrane fluorescence, and channel 1 represents the nuclei fluorescence.<br>
- y and x, which correspond to the dimensions of each slice</p>
</section>
</section>
<section id="engine" class="level2">
<h2 class="anchored" data-anchor-id="engine">Engine</h2>
<p>The engine module provides advanced functionalities for model training, including configurable training loops and evaluation functions tailored for bioinformatics applications. This module is significantly valuable when there is a need for specific workflows and pipelines that meet specific requirements. For this reason, the classes fastTrainer and visionTrainer have been created, providing tailored implementations inheriting from the Learner class.</p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L56" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="read_yaml" class="level3">
<h3 class="anchored" data-anchor-id="read_yaml">read_yaml</h3>
<blockquote class="blockquote">
<pre><code> read_yaml (yaml_path)</code></pre>
</blockquote>
<p><em>Reads a YAML file and returns its contents as a dictionary</em></p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L63" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="dictlist_to_funclist" class="level3">
<h3 class="anchored" data-anchor-id="dictlist_to_funclist">dictlist_to_funclist</h3>
<blockquote class="blockquote">
<pre><code> dictlist_to_funclist (transform_dicts)</code></pre>
</blockquote>
<p>FastTrainer is used for training models in bioinformatics applications, where specific loss functions and optimizers oriented to biological data can be used.</p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L77" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="fasttrainer" class="level3">
<h3 class="anchored" data-anchor-id="fasttrainer">fastTrainer</h3>
<blockquote class="blockquote">
<pre><code> fastTrainer (dataloaders:fastai.data.core.DataLoaders, model:&lt;built-
              infunctioncallable&gt;, loss_fn:typing.Any|None=None, optimizer
              :fastai.optimizer.Optimizer|fastai.optimizer.OptimWrapper=&lt;f
              unction Adam&gt;, lr:float|slice=0.001, splitter:&lt;built-
              infunctioncallable&gt;=&lt;function trainable_params&gt;, callbacks:f
              astai.callback.core.Callback|collections.abc.MutableSequence
              |None=None, metrics:typing.Any|collections.abc.MutableSequen
              ce|None=None, csv_log:bool=False, show_graph:bool=True,
              show_summary:bool=False, find_lr:bool=False,
              find_lr_fn=&lt;function valley&gt;,
              path:str|pathlib.Path|None=None,
              model_dir:str|pathlib.Path='models', wd:float|int|None=None,
              wd_bn_bias:bool=False, train_bn:bool=True, moms:tuple=(0.95,
              0.85, 0.95), default_cbs:bool=True)</code></pre>
</blockquote>
<p><em>A custom implementation of the FastAI Learner class for training models in bioinformatics applications.</em></p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>dataloaders</td>
<td>DataLoaders</td>
<td></td>
<td>The DataLoader objects containing training and validation datasets.</td>
</tr>
<tr class="even">
<td>model</td>
<td>callable</td>
<td></td>
<td>A callable model that will be trained on the dataset.</td>
</tr>
<tr class="odd">
<td>loss_fn</td>
<td>typing.Any | None</td>
<td>None</td>
<td>The loss function to optimize during training. If None, defaults to a suitable default.</td>
</tr>
<tr class="even">
<td>optimizer</td>
<td>fastai.optimizer.Optimizer | fastai.optimizer.OptimWrapper</td>
<td>Adam</td>
<td>The optimizer function to use. Defaults to Adam if not specified.</td>
</tr>
<tr class="odd">
<td>lr</td>
<td>float | slice</td>
<td>0.001</td>
<td>Learning rate for the optimizer. Can be a float or a slice object for learning rate scheduling.</td>
</tr>
<tr class="even">
<td>splitter</td>
<td>callable</td>
<td>trainable_params</td>
<td></td>
</tr>
<tr class="odd">
<td>callbacks</td>
<td>fastai.callback.core.Callback | collections.abc.MutableSequence | None</td>
<td>None</td>
<td>A callable that determines which parameters of the model should be updated during training.</td>
</tr>
<tr class="even">
<td>metrics</td>
<td>typing.Any | collections.abc.MutableSequence | None</td>
<td>None</td>
<td>Optional list of callback functions to customize training behavior.</td>
</tr>
<tr class="odd">
<td>csv_log</td>
<td>bool</td>
<td>False</td>
<td>Metrics to evaluate the performance of the model during training.</td>
</tr>
<tr class="even">
<td>show_graph</td>
<td>bool</td>
<td>True</td>
<td>Whether to log training history to a CSV file. If True, logs will be appended to ‘history.csv’.</td>
</tr>
<tr class="odd">
<td>show_summary</td>
<td>bool</td>
<td>False</td>
<td>The base directory where models are saved or loaded from. Defaults to None.</td>
</tr>
<tr class="even">
<td>find_lr</td>
<td>bool</td>
<td>False</td>
<td>Subdirectory within the base path where trained models are stored. Default is ‘models’.</td>
</tr>
<tr class="odd">
<td>find_lr_fn</td>
<td>function</td>
<td>valley</td>
<td>Weight decay factor for optimization. Defaults to None.</td>
</tr>
<tr class="even">
<td>path</td>
<td>str | pathlib.Path | None</td>
<td>None</td>
<td>Whether to apply weight decay to batch normalization and bias parameters.</td>
</tr>
<tr class="odd">
<td>model_dir</td>
<td>str | pathlib.Path</td>
<td>models</td>
<td>Whether to update the batch normalization statistics during training.</td>
</tr>
<tr class="even">
<td>wd</td>
<td>float | int | None</td>
<td>None</td>
<td></td>
</tr>
<tr class="odd">
<td>wd_bn_bias</td>
<td>bool</td>
<td>False</td>
<td></td>
</tr>
<tr class="even">
<td>train_bn</td>
<td>bool</td>
<td>True</td>
<td></td>
</tr>
<tr class="odd">
<td>moms</td>
<td>tuple</td>
<td>(0.95, 0.85, 0.95)</td>
<td>Tuple of tuples representing the momentum values for different layers in the model. Defaults to FastAI’s default settings if not specified.</td>
</tr>
<tr class="even">
<td>default_cbs</td>
<td>bool</td>
<td>True</td>
<td>Automatically include default callbacks such as ShowGraphCallback and CSVLogger.</td>
</tr>
</tbody>
</table>
<section id="example-train-a-model-with-configuration-from-a-yaml-file." class="level4">
<h4 class="anchored" data-anchor-id="example-train-a-model-with-configuration-from-a-yaml-file.">Example: train a model with configuration from a YAML file.</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> monai.networks.nets <span class="im">import</span> SEResNet50</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> bioMONAI.data <span class="im">import</span> BioDataLoaders</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import the data</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>image_path <span class="op">=</span> <span class="st">'_data'</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>info <span class="op">=</span> download_medmnist(<span class="st">'bloodmnist'</span>, image_path, download_only<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>batch_size <span class="op">=</span> <span class="dv">32</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>path <span class="op">=</span> Path(image_path)<span class="op">/</span><span class="st">'bloodmnist'</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>path_train <span class="op">=</span> path<span class="op">/</span><span class="st">'train'</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>path_val <span class="op">=</span> path<span class="op">/</span><span class="st">'val'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Downloading https://zenodo.org/records/10519652/files/bloodmnist.npz?download=1 to _data/bloodmnist/bloodmnist.npz
Using downloaded and verified file: _data/bloodmnist/bloodmnist.npz
Using downloaded and verified file: _data/bloodmnist/bloodmnist.npz
Saving training images to _data/bloodmnist...
Saving validation images to _data/bloodmnist...
Saving test images to _data/bloodmnist...
Removed bloodmnist.npz
Datasets downloaded to _data/bloodmnist
Dataset info for 'bloodmnist': {'python_class': 'BloodMNIST', 'description': 'The BloodMNIST is based on a dataset of individual normal cells, captured from individuals without infection, hematologic or oncologic disease and free of any pharmacologic treatment at the moment of blood collection. It contains a total of 17,092 images and is organized into 8 classes. We split the source dataset with a ratio of 7:1:2 into training, validation and test set. The source images with resolution 3×360×363 pixels are center-cropped into 3×200×200, and then resized into 3×28×28.', 'url': 'https://zenodo.org/records/10519652/files/bloodmnist.npz?download=1', 'MD5': '7053d0359d879ad8a5505303e11de1dc', 'url_64': 'https://zenodo.org/records/10519652/files/bloodmnist_64.npz?download=1', 'MD5_64': '2b94928a2ae4916078ca51e05b6b800b', 'url_128': 'https://zenodo.org/records/10519652/files/bloodmnist_128.npz?download=1', 'MD5_128': 'adace1e0ed228fccda1f39692059dd4c', 'url_224': 'https://zenodo.org/records/10519652/files/bloodmnist_224.npz?download=1', 'MD5_224': 'b718ff6835fcbdb22ba9eacccd7b2601', 'task': 'multi-class', 'label': {'0': 'basophil', '1': 'eosinophil', '2': 'erythroblast', '3': 'immature granulocytes(myelocytes, metamyelocytes and promyelocytes)', '4': 'lymphocyte', '5': 'monocyte', '6': 'neutrophil', '7': 'platelet'}, 'n_channels': 3, 'n_samples': {'train': 11959, 'val': 1712, 'test': 3421}, 'license': 'CC BY 4.0'}</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>100%|██████████| 35461855/35461855 [00:09&lt;00:00, 3590813.84it/s]
100%|██████████| 11959/11959 [00:02&lt;00:00, 4998.88it/s]
100%|██████████| 1712/1712 [00:00&lt;00:00, 5081.19it/s]
100%|██████████| 3421/3421 [00:00&lt;00:00, 5037.19it/s]</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the dataloader</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>data <span class="op">=</span> BioDataLoaders.class_from_folder(</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    path,</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    train<span class="op">=</span><span class="st">'train'</span>,</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    valid<span class="op">=</span><span class="st">'val'</span>,</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    vocab<span class="op">=</span>info[<span class="st">'label'</span>],</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    batch_tfms<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    bs<span class="op">=</span>batch_size)</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the model</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> SEResNet50(spatial_dims<span class="op">=</span><span class="dv">2</span>,</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>                   in_channels<span class="op">=</span><span class="dv">3</span>,   </span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>                   num_classes<span class="op">=</span><span class="dv">8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the trainer with configuration from a YAML file </span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>yaml_path <span class="op">=</span> <span class="st">"./data_examples/sample_config.yml"</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>trainer <span class="op">=</span> fastTrainer.from_yaml(data, model, yaml_path)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Train the model</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>trainer.fit(<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">

<style>
    /* Turns off some styling */
    progress {
        /* gets rid of default border in Firefox and Opera. */
        border: none;
        /* Needs to be in here for Safari polyfill so background images work as expected. */
        background-size: auto;
    }
    progress:not([value]), progress:not([value])::-webkit-progress-bar {
        background: repeating-linear-gradient(45deg, #7e7e7e, #7e7e7e 10px, #5c5c5c 10px, #5c5c5c 20px);
    }
    .progress-bar-interrupted, .progress-bar-interrupted::-webkit-progress-bar {
        background: #F44336;
    }
</style>
</div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">epoch</th>
<th data-quarto-table-cell-role="th">train_loss</th>
<th data-quarto-table-cell-role="th">valid_loss</th>
<th data-quarto-table-cell-role="th">accuracy</th>
<th data-quarto-table-cell-role="th">balanced_accuracy_score</th>
<th data-quarto-table-cell-role="th">precision_score</th>
<th data-quarto-table-cell-role="th">time</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>0.957569</td>
<td>0.669474</td>
<td>0.755841</td>
<td>0.682645</td>
<td>0.767054</td>
<td>00:12</td>
</tr>
</tbody>
</table>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Better model found at epoch 0 with accuracy value: 0.7558411359786987.</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="00_core_files/figure-html/cell-14-output-4.png" class="img-fluid"></p>
</div>
</div>
<p>VisionTrainer is used for computer vision applications, where image normalization or other computer vision related settings are needed.</p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L222" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
</section>
<section id="visiontrainer" class="level3">
<h3 class="anchored" data-anchor-id="visiontrainer">visionTrainer</h3>
<blockquote class="blockquote">
<pre><code> visionTrainer (dataloaders:fastai.data.core.DataLoaders, model:&lt;built-
                infunctioncallable&gt;, normalize=True, n_out=None,
                pretrained=True, weights=None,
                loss_fn:typing.Any|None=None, optimizer:fastai.optimizer.O
                ptimizer|fastai.optimizer.OptimWrapper=&lt;function Adam&gt;,
                lr:float|slice=0.001, splitter:&lt;built-
                infunctioncallable&gt;=&lt;function trainable_params&gt;, callbacks
                :fastai.callback.core.Callback|collections.abc.MutableSequ
                ence|None=None, metrics:typing.Any|collections.abc.Mutable
                Sequence|None=None, csv_log:bool=False,
                show_graph:bool=True, show_summary:bool=False,
                find_lr:bool=False, find_lr_fn=&lt;function valley&gt;,
                path:str|pathlib.Path|None=None,
                model_dir:str|pathlib.Path='models',
                wd:float|int|None=None, wd_bn_bias:bool=False,
                train_bn:bool=True, moms:tuple=(0.95, 0.85, 0.95),
                default_cbs:bool=True, cut=None, init=&lt;function
                kaiming_normal_&gt;, custom_head=None, concat_pool=True,
                pool=True, lin_ftrs=None, ps=0.5, first_bn=True,
                bn_final=False, lin_first=False, y_range=None, n_in=3)</code></pre>
</blockquote>
<p><em>Build a vision trainer from <code>dataloaders</code> and <code>model</code></em></p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>dataloaders</td>
<td>DataLoaders</td>
<td></td>
<td>The DataLoader objects containing training and validation datasets.</td>
</tr>
<tr class="even">
<td>model</td>
<td>callable</td>
<td></td>
<td>A callable model that will be trained on the dataset.</td>
</tr>
<tr class="odd">
<td>normalize</td>
<td>bool</td>
<td>True</td>
<td></td>
</tr>
<tr class="even">
<td>n_out</td>
<td>NoneType</td>
<td>None</td>
<td></td>
</tr>
<tr class="odd">
<td>pretrained</td>
<td>bool</td>
<td>True</td>
<td></td>
</tr>
<tr class="even">
<td>weights</td>
<td>NoneType</td>
<td>None</td>
<td></td>
</tr>
<tr class="odd">
<td>loss_fn</td>
<td>typing.Any | None</td>
<td>None</td>
<td>The loss function to optimize during training. If None, defaults to a suitable default.</td>
</tr>
<tr class="even">
<td>optimizer</td>
<td>fastai.optimizer.Optimizer | fastai.optimizer.OptimWrapper</td>
<td>Adam</td>
<td>The optimizer function to use. Defaults to Adam if not specified.</td>
</tr>
<tr class="odd">
<td>lr</td>
<td>float | slice</td>
<td>0.001</td>
<td>Learning rate for the optimizer. Can be a float or a slice object for learning rate scheduling.</td>
</tr>
<tr class="even">
<td>splitter</td>
<td>callable</td>
<td>trainable_params</td>
<td></td>
</tr>
<tr class="odd">
<td>callbacks</td>
<td>fastai.callback.core.Callback | collections.abc.MutableSequence | None</td>
<td>None</td>
<td>A callable that determines which parameters of the model should be updated during training.</td>
</tr>
<tr class="even">
<td>metrics</td>
<td>typing.Any | collections.abc.MutableSequence | None</td>
<td>None</td>
<td>Optional list of callback functions to customize training behavior.</td>
</tr>
<tr class="odd">
<td>csv_log</td>
<td>bool</td>
<td>False</td>
<td>Metrics to evaluate the performance of the model during training.</td>
</tr>
<tr class="even">
<td>show_graph</td>
<td>bool</td>
<td>True</td>
<td>Whether to log training history to a CSV file. If True, logs will be appended to ‘history.csv’.</td>
</tr>
<tr class="odd">
<td>show_summary</td>
<td>bool</td>
<td>False</td>
<td>The base directory where models are saved or loaded from. Defaults to None.</td>
</tr>
<tr class="even">
<td>find_lr</td>
<td>bool</td>
<td>False</td>
<td>Subdirectory within the base path where trained models are stored. Default is ‘models’.</td>
</tr>
<tr class="odd">
<td>find_lr_fn</td>
<td>function</td>
<td>valley</td>
<td>Weight decay factor for optimization. Defaults to None.</td>
</tr>
<tr class="even">
<td>path</td>
<td>str | pathlib.Path | None</td>
<td>None</td>
<td>Whether to apply weight decay to batch normalization and bias parameters.</td>
</tr>
<tr class="odd">
<td>model_dir</td>
<td>str | pathlib.Path</td>
<td>models</td>
<td>Whether to update the batch normalization statistics during training.</td>
</tr>
<tr class="even">
<td>wd</td>
<td>float | int | None</td>
<td>None</td>
<td></td>
</tr>
<tr class="odd">
<td>wd_bn_bias</td>
<td>bool</td>
<td>False</td>
<td></td>
</tr>
<tr class="even">
<td>train_bn</td>
<td>bool</td>
<td>True</td>
<td></td>
</tr>
<tr class="odd">
<td>moms</td>
<td>tuple</td>
<td>(0.95, 0.85, 0.95)</td>
<td>Tuple of tuples representing the momentum values for different layers in the model. Defaults to FastAI’s default settings if not specified.</td>
</tr>
<tr class="even">
<td>default_cbs</td>
<td>bool</td>
<td>True</td>
<td>Automatically include default callbacks such as ShowGraphCallback and CSVLogger.</td>
</tr>
<tr class="odd">
<td>cut</td>
<td>NoneType</td>
<td>None</td>
<td>model &amp; head args</td>
</tr>
<tr class="even">
<td>init</td>
<td>function</td>
<td>kaiming_normal_</td>
<td></td>
</tr>
<tr class="odd">
<td>custom_head</td>
<td>NoneType</td>
<td>None</td>
<td></td>
</tr>
<tr class="even">
<td>concat_pool</td>
<td>bool</td>
<td>True</td>
<td></td>
</tr>
<tr class="odd">
<td>pool</td>
<td>bool</td>
<td>True</td>
<td></td>
</tr>
<tr class="even">
<td>lin_ftrs</td>
<td>NoneType</td>
<td>None</td>
<td></td>
</tr>
<tr class="odd">
<td>ps</td>
<td>float</td>
<td>0.5</td>
<td></td>
</tr>
<tr class="even">
<td>first_bn</td>
<td>bool</td>
<td>True</td>
<td></td>
</tr>
<tr class="odd">
<td>bn_final</td>
<td>bool</td>
<td>False</td>
<td></td>
</tr>
<tr class="even">
<td>lin_first</td>
<td>bool</td>
<td>False</td>
<td></td>
</tr>
<tr class="odd">
<td>y_range</td>
<td>NoneType</td>
<td>None</td>
<td></td>
</tr>
<tr class="even">
<td>n_in</td>
<td>int</td>
<td>3</td>
<td></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="evaluation" class="level2">
<h2 class="anchored" data-anchor-id="evaluation">Evaluation</h2>
<p>The evaluation module provides functionalities for model evaluation, with several customizations available.</p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L363" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="display_statistics_table" class="level3">
<h3 class="anchored" data-anchor-id="display_statistics_table">display_statistics_table</h3>
<blockquote class="blockquote">
<pre><code> display_statistics_table (stats, fn_name='', as_dataframe=True)</code></pre>
</blockquote>
<p><em>Display a table of the key statistics.</em></p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L329" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="plot_histogram_and_kde" class="level3">
<h3 class="anchored" data-anchor-id="plot_histogram_and_kde">plot_histogram_and_kde</h3>
<blockquote class="blockquote">
<pre><code> plot_histogram_and_kde (data, stats, bw_method=0.3, fn_name='')</code></pre>
</blockquote>
<p><em>Plot the histogram and KDE of the data with key statistics marked.</em></p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L317" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="format_sig" class="level3">
<h3 class="anchored" data-anchor-id="format_sig">format_sig</h3>
<blockquote class="blockquote">
<pre><code> format_sig (value)</code></pre>
</blockquote>
<p><em>Format numbers with two significant digits.</em></p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L302" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="calculate_statistics" class="level3">
<h3 class="anchored" data-anchor-id="calculate_statistics">calculate_statistics</h3>
<blockquote class="blockquote">
<pre><code> calculate_statistics (data)</code></pre>
</blockquote>
<p><em>Calculate key statistics for the data.</em></p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L291" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="compute_metric" class="level3">
<h3 class="anchored" data-anchor-id="compute_metric">compute_metric</h3>
<blockquote class="blockquote">
<pre><code> compute_metric (predictions, targets, metric_fn)</code></pre>
</blockquote>
<p><em>Compute the metric for each prediction-target pair. Handles cases where metric_fn has or does not have a ‘func’ attribute.</em></p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L284" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="compute_losses" class="level3">
<h3 class="anchored" data-anchor-id="compute_losses">compute_losses</h3>
<blockquote class="blockquote">
<pre><code> compute_losses (predictions, targets, loss_fn)</code></pre>
</blockquote>
<p><em>Compute the loss for each prediction-target pair.</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> numpy.random <span class="im">import</span> standard_normal</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>a <span class="op">=</span> standard_normal(<span class="dv">1000</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>stats <span class="op">=</span> calculate_statistics(a)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>plot_histogram_and_kde(a, stats)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="00_core_files/figure-html/cell-23-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>Evaluate_model and evaluate_classification_model are two classes created in order to integrate the evaluation process in a single computation. Evaluate_model can be used on any type of task, whereas evaluate_classification_model is specifically designed for classification tasks.</p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L398" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="evaluate_model" class="level3">
<h3 class="anchored" data-anchor-id="evaluate_model">evaluate_model</h3>
<blockquote class="blockquote">
<pre><code> evaluate_model (trainer:fastai.learner.Learner,
                 test_data:fastai.data.core.DataLoaders=None, loss=None,
                 metrics=None, bw_method=0.3, show_graph=True,
                 show_table=True, show_results=True, as_dataframe=True,
                 cmap='magma')</code></pre>
</blockquote>
<p><em>Calculate and optionally plot the distribution of loss values from predictions made by the trainer on test data, with an optional table of key statistics.</em></p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>trainer</td>
<td>Learner</td>
<td></td>
<td>The model trainer object with a get_preds method.</td>
</tr>
<tr class="even">
<td>test_data</td>
<td>DataLoaders</td>
<td>None</td>
<td>DataLoader containing test data.</td>
</tr>
<tr class="odd">
<td>loss</td>
<td>NoneType</td>
<td>None</td>
<td>Loss function to evaluate prediction-target pairs.</td>
</tr>
<tr class="even">
<td>metrics</td>
<td>NoneType</td>
<td>None</td>
<td>Single metric or a list of metrics to evaluate.</td>
</tr>
<tr class="odd">
<td>bw_method</td>
<td>float</td>
<td>0.3</td>
<td>Bandwidth method for KDE.</td>
</tr>
<tr class="even">
<td>show_graph</td>
<td>bool</td>
<td>True</td>
<td>Boolean flag to show the histogram and KDE plot.</td>
</tr>
<tr class="odd">
<td>show_table</td>
<td>bool</td>
<td>True</td>
<td>Boolean flag to show the statistics table.</td>
</tr>
<tr class="even">
<td>show_results</td>
<td>bool</td>
<td>True</td>
<td>Boolean flag to show model results on test data.</td>
</tr>
<tr class="odd">
<td>as_dataframe</td>
<td>bool</td>
<td>True</td>
<td>Boolean flag to display table as a DataFrame.</td>
</tr>
<tr class="even">
<td>cmap</td>
<td>str</td>
<td>magma</td>
<td>Colormap for visualization.</td>
</tr>
</tbody>
</table>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L459" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="evaluate_classification_model" class="level3">
<h3 class="anchored" data-anchor-id="evaluate_classification_model">evaluate_classification_model</h3>
<blockquote class="blockquote">
<pre><code> evaluate_classification_model (trainer:fastai.learner.Learner,
                                test_data:fastai.data.core.DataLoaders=Non
                                e, loss_fn=None, most_confused_n:int=1,
                                normalize:bool=True, metrics=None,
                                bw_method=0.3, show_graph=True,
                                show_table=True, show_results=True,
                                as_dataframe=True, cmap=&lt;matplotlib.colors
                                .LinearSegmentedColormap object at
                                0x7fafd788fe10&gt;)</code></pre>
</blockquote>
<p><em>Evaluates a classification model by displaying results, confusion matrix, and most confused classes.</em></p>
<table class="table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>trainer</td>
<td>Learner</td>
<td></td>
<td>The trained model (learner) to evaluate.</td>
</tr>
<tr class="even">
<td>test_data</td>
<td>DataLoaders</td>
<td>None</td>
<td>DataLoader with test data for evaluation. If None, the validation dataset is used.</td>
</tr>
<tr class="odd">
<td>loss_fn</td>
<td>NoneType</td>
<td>None</td>
<td>Loss function used in the model for ClassificationInterpretation. If None, the loss function is loaded from trainer.</td>
</tr>
<tr class="even">
<td>most_confused_n</td>
<td>int</td>
<td>1</td>
<td>Number of most confused class pairs to display.</td>
</tr>
<tr class="odd">
<td>normalize</td>
<td>bool</td>
<td>True</td>
<td>Whether to normalize the confusion matrix.</td>
</tr>
<tr class="even">
<td>metrics</td>
<td>NoneType</td>
<td>None</td>
<td>Single metric or a list of metrics to evaluate.</td>
</tr>
<tr class="odd">
<td>bw_method</td>
<td>float</td>
<td>0.3</td>
<td>Bandwidth method for KDE.</td>
</tr>
<tr class="even">
<td>show_graph</td>
<td>bool</td>
<td>True</td>
<td>Boolean flag to show the histogram and KDE plot.</td>
</tr>
<tr class="odd">
<td>show_table</td>
<td>bool</td>
<td>True</td>
<td>Boolean flag to show the statistics table.</td>
</tr>
<tr class="even">
<td>show_results</td>
<td>bool</td>
<td>True</td>
<td>Boolean flag to show model results on test data.</td>
</tr>
<tr class="odd">
<td>as_dataframe</td>
<td>bool</td>
<td>True</td>
<td>Boolean flag to display table as a DataFrame.</td>
</tr>
<tr class="even">
<td>cmap</td>
<td>LinearSegmentedColormap</td>
<td>&lt;matplotlib.colors.LinearSegmentedColormap object at 0x7fafd788fe10&gt;</td>
<td>Color map for the confusion matrix plot.</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="utils" class="level2">
<h2 class="anchored" data-anchor-id="utils">Utils</h2>
<p>The utils module contains helper functions and classes to facilitate data manipulation, model setup, and training. These utilities add flexibility and convenience, supporting rapid experimentation and efficient data handling.</p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L534" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="attributesfromdict" class="level3">
<h3 class="anchored" data-anchor-id="attributesfromdict">attributesFromDict</h3>
<blockquote class="blockquote">
<pre><code> attributesFromDict (d)</code></pre>
</blockquote>
<p><em>The <a href="https://deepCLEM.github.io/bioMONAI/core.html#attributesfromdict"><code>attributesFromDict</code></a> function simplifies the conversion of dictionary keys and values into object attributes, allowing dynamic attribute creation for configuration objects. This utility is handy for initializing model or dataset configurations directly from dictionaries, improving code readability and maintainability.</em></p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L543" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_device" class="level3">
<h3 class="anchored" data-anchor-id="get_device">get_device</h3>
<blockquote class="blockquote">
<pre><code> get_device ()</code></pre>
</blockquote>
<p><em>The <a href="https://deepCLEM.github.io/bioMONAI/core.html#get_device"><code>get_device</code></a> function is used to detect if the device the code is executed in has got a CUDA-enabled GPU available. If it doesn’t, it returns CPU.</em></p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L551" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="img2float" class="level3">
<h3 class="anchored" data-anchor-id="img2float">img2float</h3>
<blockquote class="blockquote">
<pre><code> img2float (image, force_copy=False)</code></pre>
</blockquote>
<p><em>The <a href="https://deepCLEM.github.io/bioMONAI/core.html#img2float"><code>img2float</code></a> function turns an image into float representation.</em></p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L558" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="img2tensor" class="level3">
<h3 class="anchored" data-anchor-id="img2tensor">img2Tensor</h3>
<blockquote class="blockquote">
<pre><code> img2Tensor (image)</code></pre>
</blockquote>
<p><em>The <a href="https://deepCLEM.github.io/bioMONAI/core.html#img2tensor"><code>img2Tensor</code></a> function turns an image into tensor representation after turning it first into float representation.</em></p>
<hr>
<p><a href="https://github.com/deepCLEM/bioMONAI/blob/main/bioMONAI/core.py#L565" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="apply_transforms" class="level3">
<h3 class="anchored" data-anchor-id="apply_transforms">apply_transforms</h3>
<blockquote class="blockquote">
<pre><code> apply_transforms (image, transforms)</code></pre>
</blockquote>
<p><em>Apply a list of transformations, ensuring at least one is applied.</em></p>
<table class="table">
<thead>
<tr class="header">
<th></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>image</td>
<td>The image to transform</td>
</tr>
<tr class="even">
<td>transforms</td>
<td>A list of transformations to apply</td>
</tr>
</tbody>
</table>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># If we pass an empty list of transforms, it should return the input unchanged</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>test_eq(apply_transforms([<span class="dv">1</span>, <span class="dv">2</span>], []), [<span class="dv">1</span>, <span class="dv">2</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("https:\/\/deepCLEM\.github\.io\/bioMONAI");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">Copyright © 2024 Biagio Mandracchia, MIT License</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/deepCLEM/bioMONAI/">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/">
      <i class="bi bi-twitter" role="img">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>



</body></html>